pandas/io/parsers.py

```diff
@@ -1124,16 +1130,24 @@ def _validate_usecols_arg(usecols):
     msg = ("'usecols' must either be all strings, all unicode, "
            "all integers or a callable")
 
-    if usecols is not None:
-        if callable(usecols):
-            return usecols, None
-        usecols_dtype = lib.infer_dtype(usecols)
-        if usecols_dtype not in ('empty', 'integer',
-                                 'string', 'unicode'):
-            raise ValueError(msg)
+    if names is None:
+        if usecols is not None:
+            if callable(usecols):
+                return usecols, None
+            usecols_dtype = lib.infer_dtype(usecols)
+            if usecols_dtype not in ('empty', 'integer',
+                                     'string', 'unicode'):
+                raise ValueError(msg)
 
-        return set(usecols), usecols_dtype
-    return usecols, None
+            return set(usecols), usecols_dtype
+        return usecols, None
+    else:
+        missing = [c for c in usecols if c not in names]
```

Comment:
any reason if names is not None you wouldn’t also do the usecols logic?
you can’t check names if usecols is a callable; so maybe this check needs to happen separately (and after) the original

Reply:
Yeah, the names that we're passing through actually depend on the result from the first call to `_validate_usecols_arg`, so we'd have already executed that code at least once.

The logic between the first and second call differs in the C and Python parsers which is why I was concerned about getting it into one neat function, there are a few steps between before we're checking usecols against names.

I can run the original code everytime we call the function if that's preferable though? Up to you! :smile:

Reply:
yes revert to the original and give it a new name

Reply:
Are you OK with _check_usecols_in_names?

Reply:
maybe _validate_usecols_names

Reply:
Done, let me know if further changes are needed.

Thanks for the help!